contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -263,6 +259,473 @@ contract MixinMatchOrders is
+        LibFillResults.FillResults memory leftFillResults;
+        LibFillResults.FillResults memory rightFillResults;
+
+        // Loop infinitely (until broken inside of the loop), but keep a counter of how
+        // many orders have been matched.
+        for (;;) {
+            // Match the two orders that are pointed to by the left and right indices
+            LibFillResults.MatchedFillResults memory matchResults = _matchOrders(
+                leftOrder,
+                rightOrder,
+                leftSignatures[leftIdx],
+                rightSignatures[rightIdx],
+                shouldMaximallyFillOrders
+            );
+
+            // Update the orderInfo structs with the updated takerAssetFilledAmount
+            leftOrderInfo.orderTakerAssetFilledAmount = _safeAdd(
+                leftOrderInfo.orderTakerAssetFilledAmount,
+                matchResults.left.takerAssetFilledAmount
+            );
+            rightOrderInfo.orderTakerAssetFilledAmount = _safeAdd(
+                rightOrderInfo.orderTakerAssetFilledAmount,
+                matchResults.right.takerAssetFilledAmount
+            );
+
+            // Aggregate the new fill results with the previous fill results for the current orders.
+            _addFillResults(
+                leftFillResults,
+                matchResults.left
+            );
+            _addFillResults(
+                rightFillResults,
+                matchResults.right
+            );
+
+            // Update the profit in the left and right maker assets using the profits from
+            // the match.
+            batchMatchedFillResults.profitInLeftMakerAsset = _safeAdd(
+                batchMatchedFillResults.profitInLeftMakerAsset,
+                matchResults.profitInLeftMakerAsset
+            );
+            batchMatchedFillResults.profitInRightMakerAsset = _safeAdd(
+                batchMatchedFillResults.profitInRightMakerAsset,
+                matchResults.profitInRightMakerAsset
+            );
+
+            // If the leftOrder is filled, update the leftIdx, leftOrder, and leftSignature,
+            // or break out of the loop if there are no more leftOrders to match.
+            if (leftOrderInfo.orderTakerAssetFilledAmount >= leftOrder.takerAssetAmount) {
+                // Update the batched fill results once the leftIdx is updated.
+                batchMatchedFillResults.left[leftIdx++] = leftFillResults;
+                // Clear the intermediate fill results value.
+                leftFillResults = LibFillResults.FillResults(0, 0, 0, 0);
+
+                // If all of the left orders have been filled, break out of the loop.
+                // Otherwise, update the current right order.
+                if (leftIdx == leftOrders.length) {
+                    // Update the right batched fill results
+                    batchMatchedFillResults.right[rightIdx] = rightFillResults;
```

Comment:
We may want to add a note to the definition of `BatchMatchedFillResults` that there will be no fill result for orders that were not matched. Right now we label the fields as "Fill results for left/right orders", which is sort of ambiguous. A client may expect to receive fill results for all orders they pass in.

Reply:
Amir and I talked about this earlier. I don’t know that we came to a conclusion on it. I’ll talk to him in the morning and see what he thinks. I’m happy to add a comment for now, and depending on what he says I’ll either create an asana task or update the PR.

Reply:
Sounds good. For this PR, we should at least ensure there is consistent behavior across match scenarios. So we should either (i) return empty fill results for all unmatched orders, or (ii) return no empty fill results for any unmatched orders. I personally think it's okay to only return a fill result for orders that are matched.

Ex of current inconsistent behavior:
```
-- Scenario 1 --
leftOrders = [order_a, order_b]
rightOrders = [order_c]
* order_a and order_c are fully matched*
leftResult = [a_fill_result, (empty)]
rightResult = [c_fill_result]

-- Scenario 2 --
leftOrders = [order_a]
rightOrders = [order_c, order_d]
* order_a and order_c are fully matched*
leftResult = [a_fill_result]
rightResult = [c_fill_result]
```

Reply:
Note:
Chatted with Alex offline. The array initialization was updated earlier in the PR, so this is just an issue with redundant writes. We may want to refactor this section so that we only write the left/right result once, but in terms of correct output this looks good to me 👍 .